File: frontend/features/bookmarks/services/bookmarkService.ts
import { api } from "@/utils/api";

export interface BookmarkResponse {
  post_id: string;
  is_bookmarked: boolean;
  message: string;
}

export interface BookmarksListResponse {
  posts: any[];
  total_count: number;
  has_more: boolean;
  next_offset: number;
  limit: number;
}

export const bookmarkService = {
  /**
   * Toggle bookmark status for a post
   */
  async toggleBookmark(postId: string): Promise<BookmarkResponse> {
    try {
      const response = await api.post('/bookmarks', {
        post_id: postId
      });
      
      return response.data.data;
    } catch (error: any) {
      console.error('Failed to toggle bookmark:', error);
      throw new Error(error.response?.data?.error || 'Failed to toggle bookmark');
    }
  },

  /**
   * Get user's bookmarked posts with pagination
   */
  async getUserBookmarks(limit: number = 20, offset: number = 0): Promise<BookmarksListResponse> {
    try {
      const response = await api.get('/bookmarks', {
        params: { limit, offset }
      });
      
      return response.data.data;
    } catch (error: any) {
      console.error('Failed to get bookmarks:', error);
      throw new Error(error.response?.data?.error || 'Failed to get bookmarks');
    }
  },

  /**
   * Check if a post is bookmarked
   */
  async checkBookmarkStatus(postId: string): Promise<boolean> {
    try {
      const response = await api.get(`/bookmarks/status/${postId}`);
      return response.data.data.is_bookmarked;
    } catch (error: any) {
      console.error('Failed to check bookmark status:', error);
      return false; // Default to not bookmarked on error
    }
  },

  /**
   * Get bookmark status for multiple posts
   */
  async getBookmarkStatusForPosts(postIds: string[]): Promise<Record<string, boolean>> {
    try {
      // If the API doesn't support batch checking, check individually
      const results: Record<string, boolean> = {};
      
      await Promise.all(
        postIds.map(async (postId) => {
          try {
            results[postId] = await this.checkBookmarkStatus(postId);
          } catch (error) {
            results[postId] = false;
          }
        })
      );
      
      return results;
    } catch (error: any) {
      console.error('Failed to get bookmark statuses:', error);
      // Return all false on error
      return postIds.reduce((acc, postId) => {
        acc[postId] = false;
        return acc;
      }, {} as Record<string, boolean>);
    }
  }
};

File: backend/internal/services/bookmark_service.go
package services

import (
	"context"
	"fmt"

	"language-exchange/internal/models"
	"language-exchange/internal/repository"
)

type BookmarkService struct {
	bookmarkRepo repository.BookmarkRepository
	postRepo     repository.PostRepository
}

func NewBookmarkService(bookmarkRepo repository.BookmarkRepository, postRepo repository.PostRepository) *BookmarkService {
	return &BookmarkService{
		bookmarkRepo: bookmarkRepo,
		postRepo:     postRepo,
	}
}

// ToggleBookmark toggles a bookmark for a user and post
func (s *BookmarkService) ToggleBookmark(ctx context.Context, userID, postID string) (*models.BookmarkResponse, error) {
	// Verify post exists
	_, err := s.postRepo.GetByID(ctx, postID)
	if err != nil {
		return nil, fmt.Errorf("post not found: %w", err)
	}

	// Toggle bookmark
	isBookmarked, err := s.bookmarkRepo.ToggleBookmark(ctx, userID, postID)
	if err != nil {
		return nil, fmt.Errorf("failed to toggle bookmark: %w", err)
	}

	message := "Post unbookmarked"
	if isBookmarked {
		message = "Post bookmarked"
	}

	return &models.BookmarkResponse{
		PostID:       postID,
		IsBookmarked: isBookmarked,
		Message:      message,
	}, nil
}

// GetUserBookmarks retrieves all bookmarked posts for a user
func (s *BookmarkService) GetUserBookmarks(ctx context.Context, userID string, limit, offset int) ([]models.Post, int, error) {
	// Get bookmarked posts
	posts, err := s.bookmarkRepo.GetUserBookmarks(ctx, userID, limit, offset)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get user bookmarks: %w", err)
	}

	// Get total count
	totalCount, err := s.bookmarkRepo.GetUserBookmarkCount(ctx, userID)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get bookmark count: %w", err)
	}

	return posts, totalCount, nil
}

// IsBookmarked checks if a user has bookmarked a specific post
func (s *BookmarkService) IsBookmarked(ctx context.Context, userID, postID string) (bool, error) {
	return s.bookmarkRepo.IsBookmarked(ctx, userID, postID)
}

// GetBookmarkStatusForPosts returns bookmark status for multiple posts
func (s *BookmarkService) GetBookmarkStatusForPosts(ctx context.Context, userID string, postIDs []string) (map[string]bool, error) {
	return s.bookmarkRepo.GetPostBookmarkStatus(ctx, userID, postIDs)
}